File: javascript/openclaw-audit.js
#!/usr/bin/env node
// openclaw-audit.sh — portiert nach javascript
// Quelle: shell, OpenClaw@gateway1:scripts/openclaw-audit.sh
// auch in: OpenClaw@gateway2:scripts/openclaw-audit.sh
// Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

// OpenClaw read-only audit/diagnostic sweep
// Output: openclaw-audit-YYYY-MM-DD.log im selben Verzeichnis wie dieses Script

import { spawnSync } from 'child_process';
import { writeFileSync, appendFileSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';
import { hostname } from 'os';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const DATE_STAMP = new Date().toISOString().slice(0, 10);
const OUT = join(__dirname, `openclaw-audit-${DATE_STAMP}.log`);

const OC = ['openclaw', '--no-color'];

const runCommand = (command, ...args) => {
  const result = spawnSync(command, args, { encoding: 'utf8' });
  return {
    stdout: result.stdout,
    stderr: result.stderr,
    status: result.status,
    error: result.error
  };
};

const writeHeader = () => {
  const header = [
    "================================================================",
    "OpenClaw audit run",
    `Started:  ${new Date().toISOString()}`,
    `Host:     ${hostname()}`,
    `User:     ${process.env.USER || process.env.USERNAME || 'unknown'}`,
    `Version:  ${(runCommand('openclaw', '--version').stdout || 'unknown').trim()}`,
    `Output:   ${OUT}`,
    "================================================================\n"
  ].join('\n');
  
  writeFileSync(OUT, header);
};

const run_cmd = (title, ...cmd) => {
  const timestamp = new Date().toISOString();
  const commandLine = cmd.join(' ');
  
  const output = [
    "----------------------------------------------------------------",
    `### ${title}`,
    `### $ ${commandLine}`,
    `### ${timestamp}`,
    "----------------------------------------------------------------"
  ].join('\n') + '\n';
  
  appendFileSync(OUT, `\n${output}`);
  
  try {
    const result = spawnSync(cmd[0], cmd.slice(1), { 
      encoding: 'utf8',
      maxBuffer: 1024 * 1024 * 10 // 10MB buffer
    });
    
    if (result.stdout) {
      appendFileSync(OUT, result.stdout);
    }
    if (result.stderr) {
      appendFileSync(OUT, result.stderr);
    }
    
    const exitCode = result.status !== null ? result.status : (result.error ? 1 : 0);
    appendFileSync(OUT, `[exit: ${exitCode}]\n`);
  } catch (error) {
    appendFileSync(OUT, `[error: ${error.message}]\n`);
  }
};

writeHeader();

run_cmd("tasks audit --severity error", ...OC, "tasks", "audit", "--severity", "error");
run_cmd("secrets audit", ...OC, "secrets", "audit");
run_cmd("security audit", ...OC, "security", "audit");
run_cmd("plugins doctor", ...OC, "plugins", "doctor");
run_cmd("plugins deps", ...OC, "plugins", "deps");
run_cmd("plugins registry", ...OC, "plugins", "registry");
run_cmd("skills check", ...OC, "skills", "check");
run_cmd("hooks check", ...OC, "hooks", "check");
run_cmd("gateway status --deep", ...OC, "gateway", "status", "--deep");
run_cmd("channels status --probe", ...OC, "channels", "status", "--probe");
run_cmd("memory status --deep", ...OC, "memory", "status", "--deep");
run_cmd("sessions --all-agents", ...OC, "sessions", "--all-agents");
run_cmd("tasks list", ...OC, "tasks", "list");
run_cmd("cron list", ...OC, "cron", "list");

const footer = [
  "",
  "================================================================",
  `Audit complete: ${new Date().toISOString()}`,
  "================================================================"
].join('\n') + '\n';

appendFileSync(OUT, footer);

console.log(`Audit complete. Output: ${OUT}`);

File: python/openclaw-audit.py
#!/usr/bin/env python3
# openclaw-audit.sh — portiert nach python
# Quelle: shell, OpenClaw@gateway1:scripts/openclaw-audit.sh
# auch in: OpenClaw@gateway2:scripts/openclaw-audit.sh
# Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

import subprocess
import datetime
import os
import sys

def get_script_dir():
    return os.path.dirname(os.path.abspath(__file__))

def get_date_stamp():
    return datetime.datetime.now().strftime('%Y-%m-%d')

def get_output_file():
    script_dir = get_script_dir()
    date_stamp = get_date_stamp()
    return os.path.join(script_dir, f'openclaw-audit-{date_stamp}.log')

def write_header(out_file):
    with open(out_file, 'w') as f:
        f.write("================================================================\n")
        f.write("OpenClaw audit run\n")
        f.write(f"Started:  {datetime.datetime.now().isoformat()}\n")
        f.write(f"Host:     {subprocess.run(['hostname'], capture_output=True, text=True).stdout.strip()}\n")
        f.write(f"User:     {subprocess.run(['whoami'], capture_output=True, text=True).stdout.strip()}\n")
        
        try:
            version = subprocess.run(['openclaw', '--version'], capture_output=True, text=True)
            version_str = version.stdout.strip() if version.returncode == 0 else 'unknown'
        except FileNotFoundError:
            version_str = 'unknown'
            
        f.write(f"Version:  {version_str}\n")
        f.write(f"Output:   {out_file}\n")
        f.write("================================================================\n")

def run_cmd(title, cmd, out_file):
    with open(out_file, 'a') as f:
        f.write("\n")
        f.write("----------------------------------------------------------------\n")
        f.write(f"### {title}\n")
        f.write(f"### $ {' '.join(cmd)}\n")
        f.write(f"### {datetime.datetime.now().isoformat()}\n")
        f.write("----------------------------------------------------------------\n")
        
        try:
            result = subprocess.run(cmd, capture_output=True, text=True)
            f.write(result.stdout)
            if result.stderr:
                f.write(result.stderr)
            f.write(f"[exit: {result.returncode}]\n")
        except Exception as e:
            f.write(f"[error: {str(e)}]\n")

def main():
    out_file = get_output_file()
    write_header(out_file)
    
    oc_base = ['openclaw', '--no-color']
    
    commands = [
        ("tasks audit --severity error", oc_base + ['tasks', 'audit', '--severity', 'error']),
        ("secrets audit", oc_base + ['secrets', 'audit']),
        ("security audit", oc_base + ['security', 'audit']),
        ("plugins doctor", oc_base + ['plugins', 'doctor']),
        ("plugins deps", oc_base + ['plugins', 'deps']),
        ("plugins registry", oc_base + ['plugins', 'registry']),
        ("skills check", oc_base + ['skills', 'check']),
        ("hooks check", oc_base + ['hooks', 'check']),
        ("gateway status --deep", oc_base + ['gateway', 'status', '--deep']),
        ("channels status --probe", oc_base + ['channels', 'status', '--probe']),
        ("memory status --deep", oc_base + ['memory', 'status', '--deep']),
        ("sessions --all-agents", oc_base + ['sessions', '--all-agents']),
        ("tasks list", oc_base + ['tasks', 'list']),
        ("cron list", oc_base + ['cron', 'list'])
    ]
    
    for title, cmd in commands:
        run_cmd(title, cmd, out_file)
    
    with open(out_file, 'a') as f:
        f.write("\n")
        f.write("================================================================\n")
        f.write(f"Audit complete: {datetime.datetime.now().isoformat()}\n")
        f.write("================================================================\n")
    
    print(f"Audit complete. Output: {out_file}")

if __name__ == "__main__":
    main()
